native/src/fairseq2n/data/sample_data_source.cc

```diff
@@ -89,6 +93,9 @@ sample_data_source::reset(bool reset_rng)
     if (reset_rng)
         generator_.set_current_seed(seed_);
 
+    if (!allow_repeats_)
```

Comment:
>With this change, if the user calls reset, we still skip the "consumed" sampled pipelines. 

I don't think so? The way sampled pipelines are marked as "consumed" is by setting the corresponding weight for that pipeline in weight_cumsums_ to 0. So to reset, we just need to restore weight_cumsums_ to what it was originally, no? 

I don't think `test_op_works_when_allow_repeats_false_is_specified` would pass if we were still skipping consumed sampled pipelines, unless I'm not understanding what you're saying.

Reply:
Sorry, I misread this as `if (allow_repeats)`. All good. Maybe one nit optimization would be on line 52. If we allow repeats, we do not have to store the original weight cumsums since it is never needed.